equinox/_module/_module.py

```diff
@@ -0,0 +1,617 @@
+import dataclasses
+import functools as ft
+import inspect
+import itertools
+import types
+import warnings
+import weakref
+from collections.abc import Callable, Hashable
+from typing import Any, cast, ParamSpec, TypeVar
+from typing_extensions import dataclass_transform
+
+import jax
+import jax.tree_util as jtu
+import numpy as np
+from jaxtyping import Array, Bool, PyTree
+
+from .._filters import is_array, is_array_like
+from .._pretty_print import tree_pformat
+from .._tree import tree_equal
+from ._better_abstract import better_dataclass, BetterABCMeta
+from ._field import field
+
+
+# Legacy compatibibility API, passed to `strict` below.
+def StrictConfig(force_abstact: bool = False, **kwargs):
+    del kwargs
+    if force_abstact:
+        return None
+    else:
+        return False
+
+
+wrapper_field_names = {
+    "__module__",
+    "__name__",
+    "__qualname__",
+    "__doc__",
+    "__annotations__",
+}
+
+
+_abstract_module_registry = weakref.WeakSet()
+_has_dataclass_init = weakref.WeakKeyDictionary()
+_flatten_sentinel = object()
+
+
+# Used to provide a pretty repr when doing `jtu.tree_structure(SomeModule(...))`.
+@dataclasses.dataclass(slots=True)
+class _FlattenedData:
+    dynamic_field_names: tuple
+    static_fields: tuple[tuple[str, Any], ...]
+    wrapper_fields: tuple[tuple[str, Any], ...]
+
+    def __repr__(self):
+        return repr((self.dynamic_field_names, self.static_fields))[1:-1]
+
+
+class _ModuleFlattener:
+    __slots__: tuple[str, str] = ("dynamic_fs", "static_fs")
+    dynamic_fs: tuple[str, ...]
+    static_fs: tuple[str, ...]
+
+    def __init__(self, fields: tuple[dataclasses.Field[Any], ...]):
+        self.dynamic_fs = tuple(
+            [f.name for f in fields if not f.metadata.get("static", False)]
+        )
+        self.static_fs = tuple(
+            [f.name for f in fields if f.metadata.get("static", False)]
+        )
```

Comment:
Just wondering if this is faster:

```suggestion
        dynamic_fs = []
        static_fs = []
        for f in fields:
            if f.metadata.get("static", False):
                dynamic_fs.append(f.name)
            else:
                static_fs.append(f.name)
        self.dynamic_fs = tuple(dynamic_fs)
        self.static_fs = tuple(static_fs)
```

Reply:
Can be hard to tell, esp with Jax, without running some tests. And I'm in an airport about to board 😅.

Reply:
Haha! From past experience (but without having benchmarked it here) I suspect your change will actually be slightly faster.